### PR TITLE
mpich: fixed for Fujitsu compiler.

### PR DESCRIPTION
**Pull request by tkameyama**
- Fortran 77 library search bug is fixed.
- Disable float16 on fujitu compiler for compiler bug.

### CODE REVIEW

**Comment by haampie** [comment]
About the float16 config bug, can you see if this patch works? https://github.com/pmodels/mpich/pull/5031 and if so we can just add that patch unconditionally and remove the conflict rule for old versions of clang.

Also are there changes that should be contributed to mpich itself? They are happy to accept pull requests. That way we don't have to apply patches in spack all the time.

**Comment by tldahlgren** [comment]
@raffenet As a maintainer, would you care to review this PR?

**Comment by tkameyama** [comment]
> 
> 
> About the float16 config bug, can you see if this patch works? [pmodels/mpich#5031](https://github.com/pmodels/mpich/pull/5031) and if so we can just add that patch unconditionally and remove the conflict rule for old versions of clang.

The Fujitsu compiler doesn't seem to have that bug.
Fujitsu Compiler failed to compile src/mpi/coll/op/oplxor.c

> Also are there changes that should be contributed to mpich itself? They are happy to accept pull requests. That way we don't have to apply patches in spack all the time.

Yes. I will submit pull request to mpich.

File: var/spack/repos/builtin/packages/mpich/fj.3.4.patch [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+     if test -z "$len" ; then
+        len=0
+--- spack-src/autogen.sh	2021-01-21 03:04:56.000000000 +0900
++++ spack-src-fj/autogen.sh	2021-03-08 13:16:57.458375128 +0900
+@@ -371,14 +371,14 @@
+ AC_PROG_LIBTOOL
+ AC_OUTPUT
+ _EOF
+-        AUTORECONF="$autoreconf -I $libtoolm4dir"
++        AUTORECONF="$autoreconf -B $libtoolm4dir"
```
Comment (raffenet) [inline var/spack/repos/builtin/packages/mpich/fj.3.4.patch:19]: Is this compiler related or a system specific autoconf issue?

File: var/spack/repos/builtin/packages/mpich/fj.3.4.patch [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
++++            output_verbose_link_cmd='$CC $CFLAGS -v conftest.$objext 2>&1 | $GREP -v "^Configured with:" | $GREP "\-L"'
++++          ;;
++++        esac
++++        ;;
+++     esac
+++     AC_MSG_RESULT([$enable_shared])
+++ 
+diff -rNu spack-src/maint/patches/optional/confdb/fujitsu.patch spack-src-fj/maint/patches/optional/confdb/fujitsu.patch
+--- spack-src/maint/patches/optional/confdb/fujitsu.patch	1970-01-01 09:00:00.000000000 +0900
++++ spack-src-fj/maint/patches/optional/confdb/fujitsu.patch	2021-03-08 12:36:57.654873969 +0900
```
Comment (raffenet) [inline var/spack/repos/builtin/packages/mpich/fj.3.4.patch:97]: We would definitely like to have this libtool patch upstream. Is there a way we can test fujitsu compiler on our systems via container or otherwise?

File: var/spack/repos/builtin/packages/mpich/fj.3.4.patch (hunk 1 above)
Comment (tkameyama) [reply var/spack/repos/builtin/packages/mpich/fj.3.4.patch:19]: This is system specific autoconf issue.
When autoconf and libtool are installed different location (like spack),
autoreconf use libtool's libtool.m4 and don't use confdb/libtool.m4.

**Comment by tkameyama** [comment]
> > Also are there changes that should be contributed to mpich itself? They are happy to accept pull requests. That way we don't have to apply patches in spack all the time.
> 
> Yes. I will submit pull request to mpich.

submitted
pmodels/mpich#5177 pmodels/mpich#5178 pmodels/mpich#5179

**Comment by alalazo** [comment]
@raffenet @tkameyama What is the status of this PR?

**Comment by raffenet** [comment]
> @raffenet @tkameyama What is the status of this PR?

Upstreaming his patches into MPICH are blocked until a contributors agreement is signed. I don't have any issues adding them to the Spack package as an intermediate step, but the commit to patch libtool is no longer going to apply to our `main` branch.

**Comment by alalazo** [comment]
Thanks @raffenet @tkameyama I'm going to close this PR then, but feel free to resurrect it or submit a new one in case
